src/handlers/feature_handlers.cpp

```diff
@@ -4942,7 +4979,624 @@ void vehCarModelFeatureHandler::DrawShadow() {
+        {
+            lvlSegment segment;
+            lvlIntersection intersection;
+
+            float posDiffX = light->Position.X - camPosition.X;
+            float posDiffY = light->Position.Y - camPosition.Y;
+            float posDiffZ = light->Position.Z - camPosition.Z;
+
+            segment.Set(light->Position, camPosition, 0, nullptr);
+
+            if (Collide(segment, &intersection, 0, nullptr, 0x20, 0, light->Position) || Collide(segment, &intersection, 0, nullptr, 0x40, 0, light->Position))
+            {
+                posDiffX = 0.f;
+                posDiffY = 0.f;
+                posDiffZ = 0.f;
+            }
+
+            float invGlowScale = -ltLight::GlowScale;
+            light->Position.X += posDiffX * invGlowScale;
+            light->Position.Y += posDiffY * invGlowScale;
+            light->Position.Z += posDiffZ * invGlowScale;
+        }
+
+        light->DrawGlow(camPosition);
+    }
+    ltLight::DrawGlowEnd();
+
+    siren->getLensFlare()->DrawBegin();
+    for (int i = 0; i < siren->getLightCount(); i++)
+    {
+        auto light = siren->getLight(i);
+        auto lightPos = light->Position;
+        auto color = light->Color;
+
+        float intensity = light->ComputeIntensity(camPosition, 0.05f);
+        siren->getLensFlare()->Draw(lightPos, color, intensity);
+    }
+    siren->getLensFlare()->DrawEnd();
+}
+
+bool vehCarModelFeatureHandler::Collide(lvlSegment& segment, lvlIntersection* intersection, int roomId, lvlInstance* ignoreInstance, ushort instanceFlags, int collideFlags, const Vector3& lightPos)
+{
+    auto segmentInfo = segment.SegmentInfo;
+    int lightRoomId = 0;
+    if (!segmentInfo)
+    {
+        segment.SegmentInfo = (lvlSegmentInfo*)&lightRoomId;
+        lightRoomId = roomId;
+    }
+
+    auto level = *lvlLevel::Singleton;
+
+    int startRoomId = level->FindRoomId(segment.StartPos, 0);
+
+    segment.SegmentInfo->StartRoomId = startRoomId;
+
+    int endRoomId = level->FindRoomId(segment.EndPos, 0);
+
+    segment.SegmentInfo->EndRoomId = endRoomId;
+
+    intersection->IntersectionPoint.NormalizedDistance = 2.0;
+
+    intersection->Poly = 0;
+
+    auto bound = level->GetBound();
+
+    bool collide = false;
+
+    Vector3 camPosition = *(Vector3*)&gfxRenderState::sm_Camera->m30;
+
+    // gather ground and building bounds
+    if (lightPos.Dist(camPosition) < 100.f)
+        collide = (*((bool(__thiscall**)(const lvlLevelBound*, lvlSegment&, lvlIntersection*, float))*getPtr<void*>(bound, 0) + 0x14))(
+            bound,
+            segment,
+            intersection,
+            intersection->IntersectionPoint.NormalizedDistance);
+
+    auto startRoomInfo = level->GetRoomInfo(startRoomId);
+    if ((startRoomInfo->InstanceFlags & instanceFlags) != 0) // collide in start room
+    {
+        short flags;
+        for (lvlInstance* i = startRoomInfo->FirstInstance; i; i = i->GetNext())
+        {
+            *(&flags + 1) = 0;
```

Comment:
This line (and the others like it) are especially bad, since they rely on the layout of the stack